lib/ansible/modules/network/asa/asa_og.py

```diff
@@ -0,0 +1,792 @@
+
+DOCUMENTATION = """
+---
+module: asa_og
+version_added: "2.8"
+author:
+    - "Federico Olivieri (@Federico87)"
+short_description: Manage object groups on a Cisco ASA
+description:
+    - This module allows you to create and update object-group network/service on Cisco ASA device.
+extends_documentation_fragment: asa
+options:
+    name:
+        description:
+            - object-group name
+        required: true
+    group_type:
+        description:
+            - object-group type
+        choices: ['network-object', 'service-object', 'port-object']
+        required: true
+    protocol:
+        description:
+            - protocol for object-group service with port-object
+        choices: ['udp', 'tcp', 'tcp-udp']
+    host_ip:
+        description:
+            - host ip address for object-group network i.e. "network-object host 8.8.8.8"
+        type: list
+    description:
+        description:
+            - object-group description
+    group_object:
+        description:
+            - group-object for network object-group i.e. "network-object object-group ansible_test"
+        type: list
+    ip_mask:
+        description:
+            - IP address and mask for network object-group i.e. "network-object 192.168.0.0 255.255.0.0"
+        type: list
+    port_range:
+        description:
+            - port range for port-object i.e. "port-object range 0 1024"
+    port_eq:
+        description:
+            - single port for port-object i.e. "port-object eq www"
+    service_cfg:
+        description:
+            - service-object configuration protocol/direction/range,port/port
+    state:
+        description:
+            - Manage the state of the resource.
+        default: replace
+        choices: ['present', 'absent', 'replace']
+"""
+
+EXAMPLES = """
+# Note: examples below use the following provider dict to handle
+#       transport and authentication to the node.
+---
+- name: configure network object-group
+  asa_og:
+    name: ansible_test_0
+    group_type: network-object
+    state: present
+    description: ansible_test object-group description
+    host_ip: ['8.8.8.8', '8.8.4.4']
+    ip_mask: ['10.0.0.0 255.255.255.0', '192.168.0.0 255.255.0.0']
+    group_object: ['awx_lon', 'awx_ams']
+    provider: "{{ connection }}"
+  register: result
+
+- name: configure port-object object-group
+  asa_og:
+    name: ansible_test_1
+    group_type: port-object
+    state: replace
+    description: ansible_test object-group description
+    protocol: 'tcp-udp'
+    port_eq: ['1025', 'kerberos']
+    port_range: ['1025 5201', '0 1024']
+    provider: "{{ connection }}"
+  register: result
+
+- name: configure service-object object-group
+  asa_og:
+    name: ansible_test_2
+    group_type: service-object
+    state: absent
+    description: ansible_test object-group description
+    service_cfg: ['tcp destination eq 8080', 'tcp destination eq www']
+    provider: "{{ connection }}"
```

Comment:
Have you tested the implementation with connection as `network_cli`, as `local` connection shall be deprecated in future and with that the need of `provider` args too so, it's better to use `network_cli`.

Reply:
Do you have module with `netowrk_cli` implemented that I can take as example?

Reply:
@FedericoOlivieri there's not change required from code end to implement `network_cli` connection. `ASA` supports `network_cli` connection plugin. I wanted you to test the respective implementation with `network_cli` connection. Create inventory file as:
```
[asa]
<asa hostname/ip>

[asa:vars]
ansible_connection=network_cli
ansible_network_os=asa
ansible_become=yes
ansible_become_method=enable
ansible_user=<username>
ansible_password=<password>
ansible_become_pass=<en password>
```
and then test your implementation `asa_og` play without the `provider` args.

Ref: [network_cli_doc](https://docs.ansible.com/ansible/latest/plugins/connection/network_cli.html)

Reply:
Just tested and is working ok. This is a snippet of some integration test. As you can see, there is no provider.

```
{
    "_ansible_parsed": true,
    "commands": [
        "object-group service ansible_test_3 tcp-udp",
        "port-object range 1 100",
        "port-object eq talk",
        "port-object eq 65535"
    ],
    "have": [
        {
            "have_protocol": "tcp-udp",
            "have_name": [
                "tcp-udp",
                "ansible_test_3",
                "service "
            ],
            "have_group_type": "port-object",
            "have_ip_mask": [],
            "have_port_eq": [
                "www",
                "1024"
            ],
            "have_description": "th1s_IS-a_D3scrIPt10n_3xaMple-",
            "have_host_ip": [],
            "have_group_object": [],
            "have_port_range": [
                "1024 10024"
            ]
        }
    ],
    "changed": true,
    "want": [
        {
            "protocol": "tcp-udp",
            "name": "ansible_test_3",
            "port_range": [
                "1 100"
            ],
            "service_cfg": null,
            "host_ip": null,
            "group_object": null,
            "ip_mask": null,
            "state": "present",
            "port_eq": [
                "talk",
                "65535"
            ],
            "group_type": "port-object",
            "description": "th1s_IS-a_D3scrIPt10n_3xaMple-"
        }
    ],
    "invocation": {
        "module_args": {
            "authorize": null,
            "username": null,
            "passwords": null,
            "protocol": "tcp-udp",
            "service_cfg": null,
            "host_ip": null,
            "host": null,
            "password": null,
            "port": null,
            "description": "th1s_IS-a_D3scrIPt10n_3xaMple-",
            "ssh_keyfile": null,
            "name": "ansible_test_3",
            "context": null,
            "ip_mask": null,
            "auth_pass": null,
            "port_eq": [
                "talk",
                "65535"
            ],
            "port_range": [
                "1 100"
            ],
            "state": "present",
            "timeout": null,
            "provider": null,
            "group_object": null,
            "group_type": "port-object"
        }
    },
    "_ansible_no_log": false
}
```